server/lib/src/idm/server.rs

```diff
@@ -1422,85 +1425,58 @@ impl<'a> IdmServerAuthTransaction<'a> {
-                    };
-                    Ok(slock_ref)
-                }
-                None => Err(false),
-            };
+            let auth = self
+                .auth_with_unix_pass(lae.target, &lae.cleartext, ct)
+                .await?;
 
-            let maybe_slock = match maybe_slock_ref.as_ref() {
-                Ok(s) => Ok(s.lock().await),
-                Err(cred_state) => Err(cred_state),
-            };
+            match auth {
+                Some(account) => {
+                    let session_id = Uuid::new_v4();
+                    security_info!(
+                        "Starting session {} for {} {}",
+                        session_id,
+                        account.spn,
+                        account.uuid
+                    );
 
-            let maybe_valid = match maybe_slock {
-                Ok(mut slock) => {
-                    // Apply the current time.
-                    slock.apply_time_step(ct);
-                    // Now check the results
-                    if slock.is_valid() {
-                        Ok(slock)
-                    } else {
-                        Err(true)
-                    }
+                    Ok(Some(LdapBoundToken {
+                        spn: account.spn,
+                        session_id,
+                        effective_session: LdapSession::UnixBind(account.uuid),
+                    }))
                 }
-                Err(cred_state) => Err(*cred_state),
-            };
+                None => Ok(None),
+            }
+        }
+    }
 
-            match maybe_valid {
-                Ok(mut slock) => {
-                    if account
-                        .verify_unix_credential(lae.cleartext.as_str(), &self.async_tx, ct)?
-                        .is_some()
-                    {
-                        let session_id = Uuid::new_v4();
-                        security_info!(
-                            "Starting session {} for {} {}",
-                            session_id,
-                            account.spn,
-                            account.uuid
-                        );
+    pub async fn token_auth_ldap(
+        &mut self,
+        lae: &LdapTokenAuthEvent,
+        ct: Duration,
+    ) -> Result<Option<LdapBoundToken>, OperationError> {
+        match self.validate_and_parse_token_to_token(&lae.token, ct)? {
+            Token::UserAuthToken(uat) => {
+                let spn = uat.spn.clone();
+                Ok(Some(LdapBoundToken {
+                    session_id: uat.session_id,
+                    spn,
+                    effective_session: LdapSession::UserAuthToken(uat),
+                }))
+            }
+            Token::ApiToken(apit, entry) => {
+                let spn = entry
+                    .get_ava_single_proto_string(Attribute::Spn)
+                    .ok_or_else(|| {
+                        OperationError::InvalidAccountState("Missing attribute: spn".to_string())
```

Comment:
Similar nit with strings, will need to ensure import's there.

```suggestion
                        OperationError::InvalidAccountState(format!("Missing attribute: {}", Attribute::Spn))
```